src/main/java/me/kavin/piped/server/handlers/auth/UserHandlers.java

```diff
@@ -109,10 +131,247 @@ public static byte[] loginResponse(String user, String pass)
+        AuthorizationCode code = authResponse.getAuthorizationCode();
+
+        AuthorizationGrant codeGrant = new AuthorizationCodeGrant(code, callback, data.pkceVerifier);
+
+        ClientAuthentication clientAuth = new ClientSecretBasic(provider.clientID, provider.clientSecret);
+        TokenRequest tokenReq = new TokenRequest.Builder(provider.tokenUri, clientAuth, codeGrant).build();
+
+        com.nimbusds.oauth2.sdk.http.HTTPResponse tokenResponseText = tokenReq.toHTTPRequest().send();
+        OIDCTokenResponse tokenResponse = (OIDCTokenResponse) OIDCTokenResponseParser.parse(tokenResponseText);
+
+        if (!tokenResponse.indicatesSuccess()) {
+            TokenErrorResponse errorResponse = tokenResponse.toErrorResponse();
+            return HttpResponse.ofCode(500).withHtml("Failure while trying to request token:\n\n" + errorResponse.getErrorObject().getDescription());
+        }
+
+        OIDCTokenResponse successResponse = tokenResponse.toSuccessResponse();
+
+        JWT idToken = JWTParser.parse(successResponse.getOIDCTokens().getIDTokenString());
+
+        try {
+            provider.validator.validate(idToken, data.nonce);
+        } catch (BadJOSEException e) {
+            System.err.println("Invalid token received: " + e);
+            return HttpResponse.ofCode(400).withHtml("Received a bad token. Please try again");
+        } catch (JOSEException e) {
+            System.err.println("Token processing error: " + e);
+            return HttpResponse.ofCode(500).withHtml("Internal processing error. Please try again");
+        }
+
+        UserInfoRequest ur = new UserInfoRequest(provider.userinfoUri, successResponse.getOIDCTokens().getBearerAccessToken());
+        UserInfoResponse userInfoResponse = UserInfoResponse.parse(ur.toHTTPRequest().send());
+
+        if (!userInfoResponse.indicatesSuccess()) {
+            return HttpResponse.ofCode(500).withHtml(
+                    "The userinfo endpoint returned an error. Please try again or contact your oidc admin\n\n" +
+                            userInfoResponse.toErrorResponse().getErrorObject().getDescription());
+        }
+
+        UserInfo userInfo = userInfoResponse.toSuccessResponse().getUserInfo();
+
+        String sub = userInfo.getSubject().toString();
+        String sessionId;
+        try (Session s = DatabaseSessionFactory.createSession()) {
+            CriteriaBuilder cb = s.getCriteriaBuilder();
+            CriteriaQuery<OidcUserData> cr = cb.createQuery(OidcUserData.class);
+            Root<OidcUserData> root = cr.from(OidcUserData.class);
+
+            cr.select(root).where(root.get("sub").in(sub));
+
+            OidcUserData dbuser = s.createQuery(cr).uniqueResult();
+
+            if (dbuser != null) {
+                sessionId = dbuser.getUser().getSessionId();
+            } else {
+                String username = userInfo.getPreferredUsername();
+                OidcUserData newUser = new OidcUserData(sub, username, provider.name);
```

Comment:
There should be some kind username duplication check. What if two users want the same username? Potentially from different providers? :thinking: 
We likely need to autogenerate a username in that case. What if we just generate privacy friendly usernames automatically for all oidc users?

The password is defaulted to blank hash, I wonder what the [Spring's PasswordEncoder](https://docs.spring.io/spring-security/site/docs/current/api/org/springframework/security/crypto/argon2/Argon2PasswordEncoder.html) we use's behavior will be with this. I hope this won't allow to log in as any user without a password.

Reply:
> Usernames
Yes i think we already talked about this. Given that the usernames are never used when logging in with oidc, i think generating random names or just using the id is the best way. I will implement this later.

> Passwords
Given that the hash in the database is only ever compared to the output of a hash function, im pretty sure this is safe. Afaik there is no input to a hash function that outputs an empty string, so the comparison should always fail. We could also do a special value but that would just increase complexity for no reason imo.